File: apps/server-mac/PrismServer/Models/ServerConfig.swift
import Foundation

struct ServerConfig: Equatable {
    private static let lanWebBindHost = "0.0.0.0"
    private static let localAPIOriginHost = "127.0.0.1"

    var serverName: String
    var apiPort: Int
    var webPort: Int
    var discoveryEnabled: Bool
    var sessionCookieName: String
    var sessionTtlHours: Int
    var encryptionMasterKey: String
    var ollamaHost: String
    var ollamaModel: String
    var ollamaAuxiliaryModel: String
    var ollamaEmbeddingModel: String
    var qdrantURL: String
    var openAIAPIKey: String

    static let requiredLocalModels = RequiredLocalModels(
        chat: "llama3.2",
        auxiliary: "llama3.2",
        embedding: "nomic-embed-text"
    )

    static let defaults = ServerConfig(
        serverName: "Prism Server",
        apiPort: 18_787,
        webPort: 18_788,
        discoveryEnabled: true,
        sessionCookieName: "localai_session",
        sessionTtlHours: 24,
        encryptionMasterKey: "change-me-to-a-long-random-secret",
        ollamaHost: "http://localhost:11434",
        ollamaModel: "llama3.2",
        ollamaAuxiliaryModel: "llama3.2",
        ollamaEmbeddingModel: "nomic-embed-text",
        qdrantURL: "http://127.0.0.1:6333",
        openAIAPIKey: ""
    )

    func environment(applicationSupportDirectory: URL) -> [String: String] {
        var env: [String: String] = [
            "API_PORT": String(apiPort),
            "PORT": String(webPort),
            "HOSTNAME": Self.lanWebBindHost,
            "LOCALAI_API_ORIGIN": "http://\(Self.localAPIOriginHost):\(apiPort)",
            "PRISM_SERVER_NAME": serverName,
            "PRISM_DISCOVERY_ENABLED": discoveryEnabled ? "true" : "false",
            "SESSION_COOKIE_NAME": sessionCookieName,
            "SESSION_TTL_HOURS": String(sessionTtlHours),
            "ENCRYPTION_MASTER_KEY": encryptionMasterKey,
            "OLLAMA_HOST": ollamaHost,
            "OLLAMA_MODEL": ollamaModel,
            "OLLAMA_AUXILIARY_MODEL": ollamaAuxiliaryModel,
            "OLLAMA_EMBEDDING_MODEL": ollamaEmbeddingModel,
            "QDRANT_URL": qdrantURL,
            "LOCALAI_DATA_DIR": applicationSupportDirectory.appendingPathComponent("Data").path,
            "NEXT_TELEMETRY_DISABLED": "1",
            "NODE_ENV": "production"
        ]

        if !openAIAPIKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["OPENAI_API_KEY"] = openAIAPIKey
        }

        return env
    }
}

struct RequiredLocalModels: Equatable {
    let chat: String
    let auxiliary: String
    let embedding: String

    var uniqueInstallOrder: [String] {
        var seen = Set<String>()
        return [chat, auxiliary, embedding]
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .filter { seen.insert($0).inserted }
    }
}

File: apps/server-windows/core/Models/ServerConfig.cs
namespace PrismServer.Core.Models;

public sealed record ServerConfig
{
    private const string LanWebBindHost = "0.0.0.0";
    private const string LocalApiOriginHost = "127.0.0.1";

    public string ServerName { get; init; } = "Prism Server";
    public int ApiPort { get; init; } = 18_787;
    public int WebPort { get; init; } = 18_788;

    /// <summary>
    /// When false (default), Prism stays private to this PC. When true, the web
    /// and API bind to all interfaces so other devices on the network can reach
    /// it. Discovery (mDNS) only advertises while this is true.
    /// </summary>
    public bool LanAccessEnabled { get; init; } = false;
    public bool DiscoveryEnabled { get; init; } = true;
    public string SessionCookieName { get; init; } = "localai_session";
    public int SessionTtlHours { get; init; } = 24;
    public string EncryptionMasterKey { get; init; } = "change-me-to-a-long-random-secret";
    public string OllamaHost { get; init; } = "http://localhost:11434";
    public string OllamaModel { get; init; } = "llama3.2";
    public string OllamaAuxiliaryModel { get; init; } = "llama3.2";
    public string OllamaEmbeddingModel { get; init; } = "nomic-embed-text";
    public string QdrantUrl { get; init; } = "http://127.0.0.1:6333";
    public string OpenAiApiKey { get; init; } = string.Empty;

    public static ServerConfig Defaults { get; } = new();
    public static RequiredLocalModels RequiredLocalModels { get; } = new("llama3.2", "llama3.2", "nomic-embed-text");

    public Dictionary<string, string> BuildEnvironment(string applicationSupportDirectory)
    {
        var dataDirectory = Path.Combine(applicationSupportDirectory, "Data");
        var env = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["API_PORT"] = ApiPort.ToString(),
            ["PORT"] = WebPort.ToString(),
            ["WEB_PORT"] = WebPort.ToString(),
            ["PRISM_WEB_PORT"] = WebPort.ToString(),
            ["HOSTNAME"] = LanAccessEnabled ? LanWebBindHost : LocalApiOriginHost,
            ["PRISM_LAN_ACCESS"] = LanAccessEnabled ? "true" : "false",
            ["PRISM_WEB_LAN"] = LanAccessEnabled ? "1" : "0",
            ["LOCALAI_API_ORIGIN"] = $"http://{LocalApiOriginHost}:{ApiPort}",
            ["PRISM_SERVER_NAME"] = ServerName,
            ["PRISM_DISCOVERY_ENABLED"] = DiscoveryEnabled ? "true" : "false",
            ["SESSION_COOKIE_NAME"] = SessionCookieName,
            ["SESSION_TTL_HOURS"] = SessionTtlHours.ToString(),
            ["ENCRYPTION_MASTER_KEY"] = EncryptionMasterKey,
            ["OLLAMA_HOST"] = OllamaHost,
            ["OLLAMA_MODEL"] = OllamaModel,
            ["OLLAMA_AUXILIARY_MODEL"] = OllamaAuxiliaryModel,
            ["OLLAMA_EMBEDDING_MODEL"] = OllamaEmbeddingModel,
            ["QDRANT_URL"] = QdrantUrl,
            ["LOCALAI_DATA_DIR"] = dataDirectory,
            ["NEXT_TELEMETRY_DISABLED"] = "1",
            ["NODE_ENV"] = "production"
        };

        if (!string.IsNullOrWhiteSpace(OpenAiApiKey))
        {
            env["OPENAI_API_KEY"] = OpenAiApiKey;
        }

        return env;
    }
}

public sealed record RequiredLocalModels(string Chat, string Auxiliary, string Embedding)
{
    public IReadOnlyList<string> UniqueInstallOrder =>
        new[] { Chat, Auxiliary, Embedding }
            .Select(model => model.Trim())
            .Where(model => model.Length > 0)
            .Distinct(StringComparer.Ordinal)
            .ToArray();
}
